src/engines/ngspice.c

```diff
@@ -313,238 +313,56 @@ static gboolean ngspice_generate_netlist (OreganoEngine *engine, const gchar *fi
 	return TRUE;
 }
 
-/**
- * this is total bogus
- */
 static void ngspice_progress (OreganoEngine *self, double *d)
 {
 	OreganoNgSpice *ngspice = OREGANO_NGSPICE (self);
 
-	ngspice->priv->progress += 0.1;
-	if (ngspice->priv->progress > 1.)
-		ngspice->priv->progress = 1.;
-	(*d) = ngspice->priv->progress;
+	g_mutex_lock(&ngspice->priv->progress_ngspice.progress_mutex);
+	*d = ngspice->priv->progress_ngspice.progress;
+	g_mutex_unlock(&ngspice->priv->progress_ngspice.progress_mutex);
 }
 
-static void ngspice_stop (OreganoEngine *self)
+static void reader_progress (OreganoEngine *self, double *d)
 {
 	OreganoNgSpice *ngspice = OREGANO_NGSPICE (self);
-	g_io_channel_shutdown (ngspice->priv->child_iochannel, TRUE, NULL);
-	g_source_remove (ngspice->priv->child_iochannel_watch);
-	g_spawn_close_pid (ngspice->priv->child_pid);
-	g_spawn_close_pid (ngspice->priv->child_stdout);
-}
 
-/**
- * returns the number of strings in a NULL terminated array of strings
- */
-static int get_count(gchar** array) {
-	int i = 0;
-	while (array[i] != NULL)
-		i++;
-	return i;
+	g_mutex_lock(&ngspice->priv->progress_reader.progress_mutex);
+	*d = ngspice->priv->progress_reader.progress;
+	g_mutex_unlock(&ngspice->priv->progress_reader.progress_mutex);
```

Comment:
Yeah, `g_atomic_pointer_get` would be better for performance, I think. It would look like something like this:
```c
static void reader_progress (OreganoEngine *self, double *d)
{
	OreganoNgSpice *ngspice = OREGANO_NGSPICE (self);

	*d = *g_atomic_pointer_get(&ngspice->priv->progress_reader_ptr);
}
```
Instead of using an expensive mutex, a cheap atomic operation is used.

The corresponding `struct` would look like this:
```c
struct _OreganoNgSpicePriv
{
	GPid child_pid;

	Schematic *schematic;

	gboolean aborted;
	CancelInfo *cancel_info;

	GList *analysis;
	guint num_analysis;
	AnalysisTypeShared current;

	double* volatile progress_ngspice_ptr;
	double* volatile progress_reader_ptr;
};
```
The corresponding writer would look like this:
```c
void write_atomic_double(volatile void *progress_ptr_ptr, double new_val) {
	double *new_val_ptr = g_new(double, 1);
	*new_val_ptr = new_val;
	double *old_val_ptr = g_atomic_pointer_get(progress_ptr_ptr);
	g_atomic_pointer_set(progress_ptr_ptr, new_val_ptr);
	g_free(old_val_ptr);
}
```
Instead of an expensive mutex operation, we are using two cheap atomic operations and some heap allocation operations. Maybe the mutex is cheaper than the two atomic operations in the writing function.

But I did not know about atomic operations until now, so I couldn't use it.